python/examples/explore.py

```diff
@@ -0,0 +1,296 @@
+    if not os.path.isfile(local_path):
+        response = requests.get(base_url + filename)
+        with open(local_path, "wb") as file:
+            file.write(response.content)
+
+
+# sphinx_gallery_end_ignore
+
+fetch_dataset("soap_kpca_c-gap-20u.json.gz")
+chemiscope.show_input("data/soap_kpca_c-gap-20u.json.gz")
+
+
+# %%
+#
+# Example with MACE-OFF and t-SNE
+# +++++++++++++++++++++++++++++++++++++
+#
+# In this part, we are going to define another `featurize` function that runs
+# calculation of desciptors with `MACE-OFF <https://github.com/ACEsuit/mace>`_ and uses
+# `t-SNE
+# <https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html>`_ for
+# the dimensionality reduction.
+#
+# The dependencies for this example can be installed with the following command:
+#
+# .. code:: bash
+#
+#     pip install mace-torch scikit-learn
+
+# %%
+#
+# Let's import the necessary libraries.
+
+from mace.calculators import mace_off  # noqa
+from sklearn.manifold import TSNE  # noqa
+
+# %%
+#
+# Load the dataset, in our example we are reading the organic molecules.
+
+qm9_frames = ase.io.read("data/explore_qm9.xyz", ":")
+
+# %%
+#
+# Now, we are defining a `featurize` function. As on the previous example, it should
+# return the reduced data.
+
+
+def mace_off_tsne(frames):
+    # At first, we initialize a mace_off calculator:
+    descriptor_opt = {"model": "small", "device": "cpu", "default_dtype": "float64"}
+    calculator = mace_off(**descriptor_opt)
+
+    # Calculate MACE features for each frame
+    descriptors = []
+    for frame in frames:
+        structure_avg = np.mean(
+            # Only use invariant descriptors (no rotational components)
+            (calculator.get_descriptors(frame, invariants_only=True)),
+            axis=0,  # Average the descriptors over all atoms in the frame
+        )
+        descriptors.append(structure_avg)
+    descriptors = np.array(descriptors)
+
+    # Get number of jobs for parallelisation
+    n_jobs = min(len(frames), os.cpu_count())
+
+    # Apply t-SNE
+    perplexity = min(30, descriptors.shape[0] - 1)
+    reducer = TSNE(n_components=2, perplexity=perplexity, n_jobs=n_jobs)
+    return reducer.fit_transform(descriptors)
+
+
+# %%
+#
+# We can also provide the additional properties inside `chemiscope.explore`.
+# For example, let's extract dipole moment from the frames using the related helper
+# function.
+
+properties = chemiscope.extract_properties(qm9_frames, only=["mu"])
+
+# %%
+#
+# Provide the created featurizer and the properties to `chemiscope.explore`.
+
+cs = chemiscope.explore(qm9_frames, featurize=mace_off_tsne, properties=properties)
+
+# %%
+#
+# Here we display the visualisation of the pre-computed data using the described
+# function for 6k structures taken from the `QM9
+# <https://jla-gardner.github.io/load-atoms/index.html>`_ dataset.
+
+fetch_dataset("mace-off-tsne-qm9.json.gz")
+chemiscope.show_input("data/mace-off-tsne-qm9.json.gz")
```

Comment:
why is the map zoomed in here?

Reply:
here we wanted to highlight a specific cluster more closely. specifically a cluster of zwitterions which are all nicely grouped.

Reply:
Why not, this should be made clear in the accompanying text then!

Reply:
Done